Move caching of block patterns to object cache, rather than transients.

joemcgill opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->
This caches block patterns using `WP_Object_Cache` rather than transients. Caches are stored in the global group so sites with persistent caches can still get the performance benefit, while allowing this data to be cleared with a standard cache flush, while avoiding persistence issues for sites not using a persistent object cache.

Trac ticket: https://core.trac.wordpress.org/ticket/59633

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

joemcgill committed:
Move caching of block patterns to object cache, rather than transients.

This caches block patterns using `WP_Object_Cache` rather than transients. Caches are stored in the global group so sites with persistent caches can still get the performance benefit, while allowing this data to be cleared with a standard cache flush, while avoiding persistance issues for sites not using a persistent object cache.

felixarntz reviewed (commented):
@joemcgill The approach looks reasonable to me, though I wonder about not using any specific cache group. Is there existing cache usage without a specific cache group in core? I'm not opposed of committing this without a specific cache group, but would want to make sure that's conforming with how core should use the object cache functions.

I realize you may be omitting this because neither the `themes` nor the `theme_json` groups are persistent cache groups, which we therefore shouldn't use here. We could still rely on a new group (which would thus be persistent as that is the default), maybe `theme_files`?

joemcgill commented:
@felixarntz, my initial approach here is the lightest possible change—converting the transient functions to their `wp_cache_ ` alternatives. The advantage here is that we're caching this to the global caching group so this data will still be persistent on environments with object caches available, while avoiding bugs for non-persistent sites who can't easily clear transients using standard cache flushing methods. 

As is, I think this is the minimum viable solution for 6.4. However, while working on this, I'd like to open consideration of two other changes:

## Cache Groups
I think keeping this in the global group for now is fine, and we can move it to a more appropriate group later, here are the alternatives I considered:

* **Use the `themes` group:** By default this group is not persistent ([reference](https://github.com/WordPress/wordpress-develop/blob/6.4/src/wp-includes/class-wp-theme.php#L252-L264)). If this group is used, only sites who have filtered `wp_cache_themes_persistently` to `true` would receive the performance benefit, since these caches are only called once per request—when patterns are registered.
* **Use the `theme_json` group:** This group is persistent, but currently is only used for data and files related to theme.json (settings, styles, etc.). Caching patterns here as well makes the purpose of this group less clear, which is why I think it's best to avoid.
* **Add a new caching group:** Ultimately, we may want to add a new caching group here, but seems imprudent to do so during RC. I'd rather propose and get feedback on a new group structure for these types of caches during a full release cycle.

## Optional architectural improvements

WP 6.4 adds three new public methods to the `WP_Theme` to get/set/delete the pattern cache. These methods are only used in core by the private function `_wp_get_block_patterns()` which is called by the private `_register_theme_block_patterns()` on `init`. I don't like the idea of adding these public methods and needing to maintain them in the future. I think a much better solution is to move `_wp_get_block_patterns()` to a public method `WP_Theme::get_block_patters()` and make the new cache methods private. This would also be more consistent with the `WP_Theme::get_block_template_folders()` method, which is also being introduced in WP 6.4.

felixarntz reviewed (approved):
@joemcgill Thanks for clarifying. Fair points, I think we can go with the global group for now and iterate later as part of coming up with a more consistent approach in 6.5.

On a related note, regarding `theme_json`, I believe that group is non-persistent as well, so we wouldn't be able to use it anyway given we would like this data to be persistently cached for sites with an object cache.

It's also reassuring to see that all tests are still passing without any modifications 🎉

felixarntz commented:
@joemcgill Regarding the additional architectural considerations:
> WP 6.4 adds three new public methods to the `WP_Theme` to get/set/delete the pattern cache. These methods are only used in core by the private function `_wp_get_block_patterns()` which is called by the private `_register_theme_block_patterns()` on `init`. I don't like the idea of adding these public methods and needing to maintain them in the future. I think a much better solution is to move `_wp_get_block_patterns()` to a public method `WP_Theme::get_block_patters()` and make the new cache methods private. This would also be more consistent with the `WP_Theme::get_block_template_folders()` method, which is also being introduced in WP 6.4.

Generally, I agree this would be a better solution. It's late in the cycle for this kind of change, but if we can make it happen with minimal code changes, let's do it.

So we would remove `_wp_get_block_patterns()` and instead call `WP_Theme::get_block_patterns()`?

joemcgill commented:
> Generally, I agree this would be a better solution. It's late in the cycle for this kind of change, but if we can make it happen with minimal code changes, let's do it.

I agree that it’s late. The only reason I feel it’s worth doing now is because once those methods ship, they need to be maintained, so now seems better. I’ll update this PR accordingly.

joemcgill committed:
Move block patterns getter to WP_Theme.

This moves `_wp_get_block_patterns` to a new public method of the `WP_Theme` class called `get_block_patterns` and makes all of the related caching methods private to avoid future compatibility issues.

joemcgill committed:
Make WP_Theme::delete_pattern_cache public

joemcgill committed:
Update unit tests

joemcgill commented on src/wp-includes/class-wp-theme.php:
```diff
@@ -1840,6 +1798,213 @@ public function get_block_template_folders() {
+	 *    - Post Types       (comma-separated values)
+	 *    - Template Types   (comma-separated values)
+	 *
+	 * @since 6.4.0
+
+	 * @return array Block pattern data.
```
Fix missing docblock line.

```suggestion
	 * @since 6.4.0
	 *	 
	 * @return array Block pattern data.
```

joemcgill committed:
Update src/wp-includes/class-wp-theme.php

joemcgill committed:
Misc docblock and naming improvements to tests

joemcgill committed:
CS cleanup

joemcgill commented on src/wp-includes/class-wp-theme.php:
```diff
@@ -1840,6 +1798,213 @@ public function get_block_template_folders() {
 		return $this->block_template_folders;
 	}
 
+	/**
```
Reviewer note:

These methods are just moved from above, so they are organized with other block-related methods. The only changes are that the `set_pattern_cache` and `get_pattern_cache` are now private, while the `delete_pattern_cache` remains public for cache clearing purposes.

joemcgill commented on tests/phpunit/tests/blocks/wpGetBlockPatterns.php:
```diff
@@ -1,35 +1,55 @@
 <?php
```
This file name needs to be updated, but let's agree on the method name first.

felixarntz replied:
See my other comment, maybe `tests/phpunit/tests/theme/wpThemeGetBlockPatterns.php`?

felixarntz commented on tests/phpunit/tests/blocks/wpGetBlockPatterns.php:
```diff
@@ -1,35 +1,55 @@
+	 * @param WP_Theme $wp_theme A WP_Theme object.
+	 * @return array|false Returns an array of patterns if cache is found, otherwise false.
+	 */
+	public function get_pattern_cache( $wp_theme ) {
```
Nit-pick: This doesn't need to be public, so better make it private.

joemcgill replied:
Handled in [2ff2d69](https://github.com/WordPress/wordpress-develop/pull/5533/commits/2ff2d6990c4564cf12a58a9988ec623c17cd1312)

felixarntz commented on tests/phpunit/tests/blocks/wpGetBlockPatterns.php:
```diff
@@ -1,35 +1,55 @@
+ * @covers WP_Theme::get_block_patterns
  */
-class Tests_Blocks_WpGetBlockPatterns extends WP_UnitTestCase {
+class Tests_Blocks_WPThemeGetBlockPatterns extends WP_UnitTestCase {
```
Maybe better to rename this to `Tests_Theme_WPThemeGetBlockPatterns`, as this is now covering a method on the `WP_Theme` class? The test file should then be moved to the `tests/phpunit/tests/theme` directory.

IMO that makes more sense since other theme specific tests, including those related to blocks, are in that folder.

felixarntz reviewed (commented):
@joemcgill This looks great to me and pretty much good to go. A few last notes regarding your existing comments on naming.

mukeshpanchal27 commented on tests/phpunit/tests/blocks/wpGetBlockPatterns.php:
```diff
@@ -76,8 +96,8 @@ public function test_should_clear_transient_after_switching_theme() {
-			$theme2->get_pattern_cache(),
-			'The transient for block theme patterns should be set'
+			$this->get_pattern_cache( $theme2 ),
+			'The cache for block theme patterns should be set.'
```
```suggestion
			'The cache for block theme patterns should match the expected values.'
```

mukeshpanchal27 commented on tests/phpunit/tests/blocks/wpGetBlockPatterns.php:
```diff
@@ -39,33 +59,33 @@ public function test_delete_theme_cache() {
-			$theme1->get_pattern_cache(),
-			'The transient for block theme should be set'
+			$this->get_pattern_cache( $theme1 ),
+			'The cache for block theme should be set.'
```
```suggestion
			'The cache for block theme patterns should match the expected values.'
```

mukeshpanchal27 reviewed (commented):
Thank you, @joemcgill, for the PR. I have some minor suggestions. 

We should rename the file from `wpGetBlockPatterns.php` to `getBlockPatterns.php` since we didn't use the `wp_` prefix.

joemcgill committed:
Rename, move, and cleanup test class

joemcgill committed:
Apply message suggestions to assertions

joemcgill commented:
Thanks @mukeshpanchal27! I've applied your message suggestions in [c35134e](https://github.com/WordPress/wordpress-develop/pull/5533/commits/c35134e47953e42057a9958b3cc1a1197cef3b4e). I renamed and moved the test file, per @felixarntz's suggestion in [2ff2d69](https://github.com/WordPress/wordpress-develop/pull/5533/commits/2ff2d6990c4564cf12a58a9988ec623c17cd1312).

felixarntz commented:
@joemcgill Not sure why, and you probably already saw it, but it looks like unit tests are currently failing. I'm not sure what you could have changed that led to this, as they were passing yesterday and the recent changes were just about renaming?

joemcgill committed:
Fix test failures due to development mode bleed

joemcgill commented:
@felixarntz I've spent the morning tracking down why the tests started failing, and it looks like moving the file into the themes directory exposed an existing problem with our test suite, where development mode had gotten set to 'theme' (which causes the caching we're testing in this file to return `false`). I've fixed that issue in the `tests/phpunit/tests/theme/wpGetGlobalStylesheet.php` test class, and updated the test class for this functionality to be a bit more resilient against external changes in [88fa6ec](https://github.com/WordPress/wordpress-develop/pull/5533/commits/88fa6ec7cfb3e3810a234dc7dbfcb8062a346b47).

felixarntz commented on tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:
```diff
@@ -26,6 +26,9 @@ class Tests_Theme_WpGetGlobalStylesheet extends WP_Theme_UnitTestCase {
 
 	public function tear_down() {
+		// Reset development mode after each test.
+		$GLOBALS['_wp_tests_development_mode'] = '';
```
Better:
```suggestion
		unset( $GLOBALS['_wp_tests_development_mode'] );
```

The production code checks for this rely on `isset()`, not on `! empty()`, so we should rather unset any such global that may have been set during tests (as the default is unset too).

joemcgill replied:
Thanks for the suggestion. Both seem to fix the Unit test issue, but happy to apply these changes. Done in [f83c7b2](https://github.com/WordPress/wordpress-develop/pull/5533/commits/f83c7b23a659725fe64606a6270f55286882db66). In the future we should consider moving all of these types of things to a test utility function.

felixarntz commented on tests/phpunit/tests/theme/wpThemeGetBlockPatterns.php:
```diff
@@ -0,0 +1,198 @@
+ */
+class Tests_Theme_WPThemeGetBlockPatterns extends WP_UnitTestCase {
+
+	public static function wpSetUpBeforeClass() {
+		// Ensure development mode is reset before running these tests.
+		$GLOBALS['_wp_tests_development_mode'] = '';
+	}
+
+	public static function wpTearDownAfterClass() {
+		// Ensure development mode is reset after running these tests.
+		$GLOBALS['_wp_tests_development_mode'] = '';
+	}
```
See above:
```suggestion
	public static function wpSetUpBeforeClass() {
		// Ensure development mode is reset before running these tests.
		unset( $GLOBALS['_wp_tests_development_mode'] );
	}

	public static function wpTearDownAfterClass() {
		// Ensure development mode is reset after running these tests.
		unset( $GLOBALS['_wp_tests_development_mode'] );
	}
```

felixarntz reviewed (approved):
Thanks @joemcgill, that makes sense. I had run into a similar situation the other day when working on #5467, I should have considered to fix this holistically 🙃 

One point of feedback below, we should rather unset the global than setting it to an empty string to clean up global state, but that's only a minor issue that could be fixed pre-commit.

joemcgill committed:
Unset _wp_tests_development_mode global

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

joemcgill commented:
Merged in https://core.trac.wordpress.org/changeset/56978
